python/ray/util/client/dataclient.py

```diff
@@ -38,19 +40,20 @@ def __init__(self, client_worker: "Worker", client_id: str, metadata: list):
         self._metadata = metadata
         self.data_thread = self._start_datathread()
 
-        # Track outstanding requests to resend in case of disconnection
+        # Track outstanding requests to resend in case of disconnection.
+        # Maps request serial number to request.
         self.outstanding_requests: Dict[int, Any] = OrderedDict()
 
         # Serialize access to all mutable internal states: self.request_queue,
-        # self.ready_data, self.asyncio_waiting_data,
-        # self._in_shutdown, self._req_id, self.outstanding_requests and
-        # calling self._next_id()
+        # self.ready_data, self.asyncio_waiting_data, self._in_shutdown and
+        # self.outstanding_requests.
         self.lock = threading.Lock()
 
         # Waiting for response or shutdown.
         self.cv = threading.Condition(lock=self.lock)
 
         self.request_queue = queue.Queue()
+        # Maps request ID (Python object address) to response.
```

Comment:
Why are we switching from `request.req_id` to `id(request)`

Reply:
Ah okay, I guess it's since req_id generation is now deferred depending on how far along request_queue is processed

Reply:
Exactly, this enables not touching req_id in async send.